core/src/main/java/org/apache/sedona/core/formatMapper/shapefileParser/parseUtils/shp/ShapeSerde.java

```diff
@@ -1,264 +0,0 @@
- * "License"); you may not use this file except in compliance
- * with the License.  You may obtain a copy of the License at
- *
- *   http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing,
- * software distributed under the License is distributed on an
- * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
- * KIND, either express or implied.  See the License for the
- * specific language governing permissions and limitations
- * under the License.
- */
-
-package org.apache.sedona.core.formatMapper.shapefileParser.parseUtils.shp;
-
-import com.esotericsoftware.kryo.io.Input;
-import org.locationtech.jts.geom.Geometry;
-import org.locationtech.jts.geom.GeometryFactory;
-import org.locationtech.jts.geom.LineString;
-import org.locationtech.jts.geom.MultiLineString;
-import org.locationtech.jts.geom.MultiPoint;
-import org.locationtech.jts.geom.MultiPolygon;
-import org.locationtech.jts.geom.Point;
-import org.locationtech.jts.geom.Polygon;
-
-import java.nio.ByteBuffer;
-import java.nio.ByteOrder;
-
-/**
- * Provides methods to efficiently serialize and deserialize geometry types
- * using shapefile format developed by ESRI. Does not serialize user data
- * attached to the geometry.
- * <p>
- * Supports Point, LineString, Polygon, MultiPoint, MultiLineString and
- * MultiPolygon types.
- * <p>
- * Compatible with the family of {@link ShapeReader} classes.
- * <p>
- * First byte contains {@link ShapeType#id}. The rest is type specific.
- * Point: 8 bytes for X coordinate, followed by 8 bytes for Y coordinate.
- * LineString is serialized as MultiLineString.
- * MultiLineString: 16 bytes for envelope, 4 bytes for the number of line strings,
- * 4 bytes for total number of vertexes, 16 * num-vertexes for
- * XY coordinates of all the vertexes.
- * Polygons is serialized as MultiPolygon.
- * MultiPolygon: 16 bytes for envelope, 4 bytes for the total number of exterior and
- * interior rings of all polygons, 4 bytes for total number of vertexes,
- * 16 * num-vertexes for XY coordinates of all the vertexes. The vertexes
- * are written one polygon at a time, exterior ring first, followed by
- * interior rings.
- */
-public class ShapeSerde
-{
-    private static final int POINT_LENGTH = 1 + 2 * ShapeFileConst.DOUBLE_LENGTH;
-
-    public static byte[] serialize(Geometry geometry)
-    {
-        if (geometry instanceof Point) {
-            return serialize((Point) geometry);
-        }
-
-        if (geometry instanceof MultiPoint) {
-            return serialize((MultiPoint) geometry);
-        }
-
-        if (geometry instanceof LineString) {
-            return serialize((LineString) geometry);
-        }
-
-        if (geometry instanceof MultiLineString) {
-            return serialize((MultiLineString) geometry);
-        }
-
-        if (geometry instanceof Polygon) {
-            return serialize((Polygon) geometry);
-        }
-
-        if (geometry instanceof MultiPolygon) {
```

Comment:
A few issues with the our old ShapeSerde serializer:

1. Does not serializer/deserialize GeometryCollection. A GeometryCollection object should be decomposed to its internal objects and serialize
2. Does not serialize/deserialize Multi-Multi object. It currently allows one level of nested objects

Reply:
1. I think that the old serde handles GeometryCollection. we also have a test for that - [GeometrySerdeTest](https://github.com/apache/incubator-sedona/blob/4800676929d3d07db2f7e6421a643ffd562f90a3/core/src/test/java/org/apache/sedona/core/formatMapper/shapefileParser/shapes/GeometrySerdeTest.java#L55)

This is handled at the write method of the [GeometrySerde](https://github.com/apache/incubator-sedona/blob/2e60fc07b0eae78ccae3876d970e677fc9319c40/core/src/main/java/org/apache/sedona/core/geometryObjects/GeometrySerde.java#L72)

2. What do you mean by the multi-multi object? For example, a multi polygon can have a nested polygons hierarchy? do you have some examples for that?

Reply:
@netanel246 I remember there was a GitHub issue that you have a multipolygon inside a multipolygon. But I couldn't find it. https://docs.google.com/spreadsheets/d/e/2PACX-1vQ4G8zwdfVhvoJAMWmBAGWNDTtE32Oh0dk95BF5HL996k0YCRDyU0cXry1xwuTGcQ4sixB0NTc4K4-0/pubhtml

1.Anyway, would you please make WKB serializer as an optional serializer? So if the old serializer does not support some very special geometries, the user could use the WKB serializer.
2. Would you please also make the SQL module serializer also allows both the old Serde and WKB serializer?

Thanks,
Jia

Reply:
I saw [here](https://github.com/geotools/geotools/blob/main/modules/plugin/shapefile/src/main/java/org/geotools/data/shapefile/shp/ShapefileReader.java) that geotools already implemented Shapefile reader and writer.

I saw the source code and it seems that they already handle holes in polygons (the bug that cause us to move to WKB: [ST_GeomFromWkt and ST_GeomFromText converts multipolygon to polygon](https://docs.google.com/spreadsheets/d/e/2PACX-1vQ4G8zwdfVhvoJAMWmBAGWNDTtE32Oh0dk95BF5HL996k0YCRDyU0cXry1xwuTGcQ4sixB0NTc4K4-0/pubhtml))

I wonder if we can use that as a SerDe instead of creating two serdes that might be confusing for users.
What do you think?

Reply:
@netanel246 I wouldn't recommend GeoTools Serde due to their license issue. According to ASF policy, components that require Category X license has to be an optional feature. Serializer is the fundamental component in Sedona. If we use their library, I am afraid that we won't be able to graduate from the incubator.

Reply:
OK.
So I will add support for both of the serdes.